library/mps_writer.h

```diff
@@ -0,0 +1,377 @@
+ * mbedtls_mps_writer_get(), while previously obtained output buffers can be
+ * marked processed through calls to mbedtls_mps_writer_commit(), emptying the
+ * list of output buffers and invalidating them from the consumer's perspective.
+ *
+ */
+
+#ifndef MBEDTLS_MPS_WRITER_H
+#define MBEDTLS_MPS_WRITER_H
+
+#include <stdio.h>
+#include <stdint.h>
+
+#include <stdio.h>
+
+#include "mps_common.h"
+#include "mps_error.h"
+
+struct mbedtls_mps_writer;
+typedef struct mbedtls_mps_writer mbedtls_mps_writer;
+
+/** \brief The type of states for the reader.
+ *
+ *  Possible values are:
+ *  - #MBEDTLS_MPS_WRITER_PROVIDING (initial state)
+ *    The writer awaits buffers for holding outgoing
+ *    data to be assigned to it via mbedtls_writer_feed().
+ *  - #MBEDTLS_MPS_WRITER_CONSUMING
+ *    The writer has buffers to serve write requests from.
+ **/
+typedef unsigned char mbedtls_mps_writer_state_t;
+#define MBEDTLS_MPS_WRITER_PROVIDING ( (mbedtls_mps_writer_state_t) 0)
+#define MBEDTLS_MPS_WRITER_CONSUMING ( (mbedtls_mps_writer_state_t) 1)
+
+struct mbedtls_mps_writer
+{
+    /** The current buffer to hold outgoing data.      */
+    unsigned char *out;
+    /** The queue buffer from which to serve write requests that would
+     *  exceed the current outgoing data buffer's bounds. May be \c NULL. */
+    unsigned char *queue;
+    /** The size in bytes of the outgoing data buffer \c out. */
+    mbedtls_mps_stored_size_t out_len;
+    /** The length of the queue buffer \c queue. */
+    mbedtls_mps_stored_size_t queue_len;
+
+    /** The offset from the beginning of the outgoing data buffer indicating
+     *  the amount of data that the user has already finished writing.
+     *
+     *  Note: When a queue buffer is in use, this may be larger than the length
+     *        of the outgoing data buffer, and is computed as if the outgoing
+     *        data buffer was immediately followed by the queue buffer.
+     *
+     * This is only used when the writer is in consuming state, i.e.
+     * <code>state == MBEDTLS_MPS_WRITER_CONSUMING</code>; in this case, its value
+     * is smaller or equal to <code>out_len + queue_len</code>.
+     */
+    mbedtls_mps_stored_size_t committed;
+
+    /** The offset from the beginning of the outgoing data buffer of the
+     *  end of the last fragment handed to the user.
+     *
+     *  Note: When a queue buffer is in use, this may be larger than the
+     *  length of the outgoing data buffer, and is computed as if the outgoing
+     *  data buffer was immediately followed by the queue buffer.
+     *
+     *  This is only used when the writer is in consuming state,
+     *  i.e. <code>state == MBEDTLS_MPS_WRITER_CONSUMING</code>; in this case,
+     *  its value is smaller or equal to <code>out_len + queue_len</code>.
+     */
+    mbedtls_mps_stored_size_t end;
+
+    /** In consuming state, this denotes the size of the overlap between the
```

Comment:
Similar to the reader, the meaning of this field depends on whether the writer is producing or consuming.

In producing, it's the amount of data copied into the queue buffer.

In consuming, it is indeed the amount of overlap between queue and output buffer, or 0 if no queue is used.

In the reader, I used the `acc_share` union for the overlap. I suggest to do the same here, as I agree that unconditionally naming it `queue_next` (but loading it as `queue_overlap` in many places) is confusing.

Ok?

Reply:
Yes with the union we can have two names which will make things easier to understand. Otherwise, in the case of the CONSUMING state wouldn't it be better to keep in the context not the overlap but rather the actual used size of the output buffer. Thus rather keeping `outlen - overlap` rather than the overlap. Looking at the code it seems that but in one case the overlap is always used as`outlen-overlap`. To me it would also make things easier to reason about.